kingpin/actors/spotinst.py

```diff
@@ -0,0 +1,493 @@
+        validation of the supplied ElastiGroup config. It cannot verify that
+        instances will truly launch, but it can help catch obvious errors.
+
+        It does require authentication, which is sad.
+        """
+        yield self._client.aws.ec2.validate_group.http_post(
+            group=self._config['group'])
+
+    @gen.coroutine
+    def _precache(self):
+        """Pre-populate a bunch of data.
+
+        Searches for the list of ElastiGroups and stores the existing
+        configuration for an ElastiGroup if it matches the name of the one
+        we're managing here.
+
+        Attempts light schema-validation of the desired ElastiGroup config to
+        try to catch errors early in the Dry run.
+        """
+        self._group = yield self._get_group()
+        yield self._validate_group()
+
+    @gen.coroutine
+    def _get_state(self):
+        if self._group:
+            raise gen.Return('present')
+
+        raise gen.Return('absent')
+
+    @gen.coroutine
+    @dry('Would have created ElastiGroup')
+    def _set_state(self):
+        self.log.info('Creating ElastiGroup %s' % self.option('name'))
+        yield self._client.aws.ec2.create_group.http_post(
+            group=self._config['group'])
+
+    @gen.coroutine
+    def _compare_config(self):
+        new = copy.deepcopy(self._config)
+        existing = copy.deepcopy(self._group)
+
+        # Strip out some of the Spotinst generated and managed fields that
+        # should never end up in either our new or existing configs.
+        for field in ('id', 'createdAt', 'updatedAt'):
+            new['group'].pop(field, None)
+            existing['group'].pop(field, None)
+
+        # We only allow a user to supply a single subnetId for each AZ (this is
+        # handled by the ElastiGroupSchema). Spotinst returns back though both
+        # the original setting, as well as a list of subnetIds. We purge that
+        # from our comparison here.
+        for az in existing['group']['compute']['availabilityZones']:
+            az.pop('subnetIds', None)
+
+        diff = utils.diff_dicts(existing, new)
+
+        if diff:
+            self.log.warning('Group configurations do not match')
+            for line in diff.split('\n'):
+                self.log.info('Diff: %s' % line)
+            return False
+
+        return True
+
+    @gen.coroutine
+    def _get_config(self):
+        """Not really used, but a stub for correctness"""
+        raise gen.Return(self._group)
+
+    @gen.coroutine
+    @dry('Would have updated ElastiGroup config')
+    def _set_config(self):
+        group_id = self._group['group']['id']
+        self.log.info('Updating ElastiGroup %s' % group_id)
+
+        # There are certain fields that simply cannot be updated -- strip them
+        # out. We have a warning up in the above _compare_config() section that
+        # will tell the user about this in a dry run.
+        if 'capacity' in self._config['group']:
```

Comment:
@mikhail, is there a more clever way for me to create a list of nested field names and access them in a loop? Something like:
```
fields = ( 'capacity.unit', 'compute.product' )
for field in fields:
  if not self._config[field] == self._group[field]:
    del self._config[field]
```

?